pkg/fsm/internal/reconciler.go

```diff
@@ -250,16 +249,27 @@ func (r *fsmReconciler[T, Obj]) reconcile(
 			r.metrics.RecordStateDuration(typedObjectRef.GroupVersionKind(), currentState.Name, time.Since(start))
 
 			condition.LastTransitionTime = metav1.Now() // set status condition last transition time
-			condition.Status = corev1.ConditionTrue     // set status condition to true if state is done
+			condition.Status = corev1.ConditionTrue     // default status condition to true if state is done
+
+			if result.RequeueAfterCompletion {
+				// the last Result type with RequeueAfterCompletion==true takes precedence
+				result.RequeueAfterCompletionState = currentState.Name
+				requeueAfterCompletion = result
+			}
 
-			if !result.IsDone() {
+			// mark the state's condition as failed if not done or the state signals a requeue after FSM completion
+			if !result.IsDone() || result.RequeueAfterCompletion {
 				// falsify condition if provided, set message and reason
 				if !condition.IsEmpty() {
 					condition.Status = corev1.ConditionFalse
 					condition.Message, condition.Reason = result.GetMessageAndReason()
 					conditions.SetConditions(condition)
 				}
 				return obj, conditions, result.WrapError(fmt.Sprintf("transitioning state %q", currentState.Name))
```

Comment:
if I return a requeue after completion, won't the return on line 268 end the loop early?

Reply:
Nice catch! I've fixed this and added a test case guarding against it